docs/architecture/017-applicant-ux.md

```diff
@@ -0,0 +1,89 @@
+
+This can be irritating if you're working on a long application, and
+just want to save your work and go home. It's less of an issue if
+users save early & often.
+
+**Possible solution**: Show validation errors to the user, but perform
+the save anyway. Show errors in yellow?
+
+Historical note: historically (in the pre-SPA days) REMS used to work
+like this. This topic was more recently touched on in
+[#2117](https://github.com/CSCfi/rems/issues/2117). Interestingly
+enough, the idea idea seems to have been to only perform schema-style
+validations on drafts. However, the email content validation was
+included in the spec, so subsequent field types added their content
+validations to save-draft (see e.g.
+[#2567](https://github.com/CSCfi/rems/pull/2567),
+[#2594](https://github.com/CSCfi/rems/pull/2594))
+
+Idea: Think of REMS validations more like spellchecking and less like
+validations. Higlight in red things that the applicant will want to
+fix, but don't prevent saving.
+
+## B) Validation errors are shown late
+
+You need to perform an action (save/submit) to see the validation
+errors. Technically we should be able to run the validation code on
+the frontend at any point. There's no need to talk to the backend.
+
+**Possible solution**: Validate when user stops typing / leaves field.
+
+Questions:
+- What about partially filled in forms? It would be best to not show validation errors for fields that haven't been edited yet.
+- What if a user navigates away from the application and back? Do they see the validation errors, or perhaps just the state of the application as-is before they write something.
+- What if a user starts filling in field 3? Do they see validation errors for fields 1-3 or only 3?
+- What to do with required fields? Should they all turn red immediately or only upon save? Or only upon submit?
+
+Idea: Three states for application fields:
+- untouched (not filled in at all)
+- edited but not valid (can save, can't submit)
+- filled in and completely valid (can submit)
+
+Idea: Only show the validation errors for the fields the user actively changes in that "session". Nothing validated when the page is opened.
+
+## C) You need to save your work explicitly
+
+You need to click on the "Save draft" button to save your work.
+Arguably not a problem, users are used to Save buttons? What if REMS
+automatically saved your work (a la Google Docs)?
+
+**Possible solution 1**: Autosave (cache) into browser localstorage We
+wouldn't call this caching "saving" to the user, instead their work
+would automatically be preserved in their browser, so if they e.g.
+navigate away from the application page and navigate back, they'll
+still see their work-in-progress application.
+
+**Possible solution 2**: Autosave into REMS db. Also needs problem A
+to be fixed. Cons: lots of events generated to the REMS db. Could work
+around this by implementing draft saving (or perhaps just auto-saving)
+separately from the application events.
```

Comment:
depending on what events can occur during draft phase, we could replace the current head of the event stack if the head has `eventdata->>'event/type' = 'application.event/draft-saved'`. this would leave interleaved draft-saved events depending on encasing events, but would skip a tremendous amount of extra events in case of auto-save

Reply:
I am sorry, I do not quite get what you are trying to do here, can you maybe provide a bit more details?

Reply:
@aatkin yes that's one of the solutions. However,  we have been reluctant with modifying the event history and this would be a step that breaks that promise.